activerecord/lib/active_record/reflection.rb

```diff
@@ -516,6 +516,14 @@ def initialize(name, scope, options, active_record)
         @foreign_key = nil
         @association_foreign_key = nil
         @association_primary_key = nil
+        if options[:query_constraints]
+          ActiveRecord.deprecator.warn <<~MSG.squish
+            Setting `query_constraints:` option on `#{active_record}.#{macro} :#{name}` is deprecated
+            and its behavior may be modifed in Rails 8.
+            To preserve the current behavior please use the `foreign_key` option instead.
```

Comment:
I'm not sure exactly why but the text here feels confusing. It's not immediately clear that foreign key is a replacement for query constraints. The implication that it "might" be different in the future isn't clear. Maybe something more like:

```ruby
Setting `query_constraints:` option on `#{active_record}.#{macro} :#{name}` is deprecated.
To maintain current behavior, use the `foreign_key` option instead.
```

Reply:
Yeah so I initially came up with a straightforward message like you suggested but then I tried to convey that `query_constraints:` is not going away, there will just be new behavior which may actually fit your application better but at the same time we can't promise anything until we release the actual new behavior.

So I'll apply your suggestion, it's better to be straightforward as `foreign_key` being a direct substitution is what we can guarantee at the moment.

Thanks!